File: apps/operations-web/src/api/intelligence.ts
import { useQuery } from "@tanstack/react-query";

import { apiRequest } from "./client";

/**
 * Owner intelligence (D-089). Every figure below is computed by the API from the canonical
 * invoices, ledgers and receipts; the client only displays it. Amounts are decimal strings that
 * always travel with their currency, lists are grouped per currency and never summed together.
 * Scores and bands are workflow rankings, never probabilities; cash-flow has no forecast field.
 */
export type Scalar = string | number | boolean | null;
export interface Reason { code: string; value: Scalar; context: Record<string, Scalar> }
export type PriorityBand = "HIGH" | "MEDIUM" | "LOW";
export type InactivityStatus = "NORMAL" | "WATCH" | "AT_RISK" | "LAPSED" | "INSUFFICIENT_HISTORY";

export interface PriorityItem {
  customer_id: string;
  customer_name: string;
  customer_grade: string | null;
  currency: string;
  score: number;
  band: PriorityBand;
  components: Record<"collection_urgency" | "relationship_inactivity" | "activity_decline" | "friction_signals", number | null>;
  reasons: Reason[];
  suggested_action_code: string;
  inactivity_status: InactivityStatus;
  outstanding_balance: string;
  days_since_last_purchase: number | null;
}
export interface PrioritiesResponse { as_of: string; groups: { currency: string; items: PriorityItem[] }[] }

export interface InactivityItem {
  customer_id: string;
  customer_name: string;
  customer_grade: string | null;
  currency: string;
  status: InactivityStatus;
  recency_ratio: string | null;
  days_since_last_purchase: number | null;
  median_purchase_interval_days: string | null;
  purchase_interval_mad_days: string | null;
  invoice_count_lifetime: number;
  invoice_count_90d: number;
  sales_30d: string;
  sales_90d: string;
  outstanding_balance: string;
  last_purchase_at: string | null;
  reason_codes: string[];
}
export interface InactivityResponse { as_of: string; groups: { currency: string; items: InactivityItem[] }[] }

export interface AnomalyItem {
  type: string;
  severity: "HIGH" | "WATCH";
  detected_at: string;
  subject_type: "TENANT" | "CUSTOMER" | "INVOICE";
  subject_id: string | null;
  metric: string;
  observed_value: string;
  baseline: { method: string; median: string | null; mad: string | null; sample_size: number | null };
  reason_code: string;
  details: Record<string, Scalar>;
}
export interface AnomaliesResponse {
  as_of: string;
  window: { start: string; end: string; timezone: string; block_days: number; baseline_blocks: number };
  groups: { currency: string; items: AnomalyItem[]; insufficient_history: string[] }[];
}

export interface CurrencyCashFlow {
  currency: string;
  position: { customer_receivables: string; customer_credit: string; overdue_receivables: string; overdue_customer_count: number; supplier_payables: string; supplier_credit: string };
  ageing: { bucket: string; amount: string; customer_count: number }[];
  historical_flow: { customer_receipts: string; customer_refunds: string; supplier_payments: string; net_customer_collections: string; average_weekly_collections: string | null };
  planned_collections: { amount: string; task_count: number; from_date: string; through_date: string; source: string; projection_warning_code: string };
}
export type CashFlowPeriod = "30d" | "90d" | "1y" | "all";
export interface CashFlowResponse {
  as_of: string;
  period: { key: string; start: string | null; end: string; timezone: string };
  overdue_threshold_days: number | null;
  currencies: CurrencyCashFlow[];
}

export interface CopilotStatus { configured: boolean; provider: string | null; model: string | null }
/** What every written answer carries, from the assistant or a contextual explanation. */
export interface GroundedAnswer {
  answer: string;
  references: { ref: string; customer_id: string; customer_name: string }[];
  grounding: { tool: string; period: string | null; currency: string | null; ok: boolean }[];
  warnings: string[];
  unverified_numbers: string[];
}
export interface CopilotTurn { role: "user" | "assistant"; content: string }
export interface CopilotResponse extends GroundedAnswer {
  conversation_id: string;
  /** The answer as the provider wrote it (customer references, not names): the history to send back. */
  conversation_text: string;
}

/**
 * Contextual explanations (D-091): wording of facts the server assembles for one customer, one
 * unusual change or the cash position. An anomaly is named as the screen shows it (currency,
 * position, type, subject); the server re-checks it and answers 409 ANOMALY_CHANGED otherwise.
 */
export type ExplainContext =
  | { kind: "customer"; customer_id: string }
  | { kind: "anomaly"; currency: string; index: number; type: string; subject_id: string | null }
  | { kind: "cash"; period: CashFlowPeriod; currency?: string | null };
export interface ExplanationResponse extends GroundedAnswer { kind: ExplainContext["kind"]; as_of: string }

/** Counted anomaly metrics; every other anomaly metric is an amount in the group's currency. */
export const COUNT_METRICS: ReadonlySet<string> = new Set(["cancelled_invoices_7d", "receipt_reversals_7d"]);

// Intelligence is computed on request over the whole business, so one answer is shared by the
// screens that show it (Work, Customers, a customer's record) for a minute instead of recomputed.
const SHARED = { staleTime: 60_000, retry: 1 } as const;

export const intelligenceKeys = {
  all: (tenantId: string) => ["intelligence", tenantId] as const,
};

export function usePriorities(tenantId: string, enabled = true) {
  return useQuery({
    queryKey: [...intelligenceKeys.all(tenantId), "priorities"],
    queryFn: () => apiRequest<PrioritiesResponse>(`/api/v1/intelligence/priorities?tenant_id=${tenantId}&limit=200`),
    enabled,
    ...SHARED,
  });
}

export function useInactivity(tenantId: string, enabled = true) {
  return useQuery({
    queryKey: [...intelligenceKeys.all(tenantId), "inactivity"],
    queryFn: () => apiRequest<InactivityResponse>(`/api/v1/intelligence/inactivity?tenant_id=${tenantId}&limit=500`),
    enabled,
    ...SHARED,
  });
}

export function useAnomalies(tenantId: string) {
  return useQuery({
    queryKey: [...intelligenceKeys.all(tenantId), "anomalies"],
    queryFn: () => apiRequest<AnomaliesResponse>(`/api/v1/intelligence/anomalies?tenant_id=${tenantId}`),
    ...SHARED,
  });
}

export function useCashFlow(tenantId: string, period: CashFlowPeriod) {
  return useQuery({
    queryKey: [...intelligenceKeys.all(tenantId), "cash-flow", period],
    queryFn: () => apiRequest<CashFlowResponse>(`/api/v1/intelligence/cash-flow?tenant_id=${tenantId}&period=${period}`),
    ...SHARED,
  });
}

export function useCopilotStatus(tenantId: string) {
  return useQuery({
    queryKey: [...intelligenceKeys.all(tenantId), "copilot-status"],
    queryFn: () => apiRequest<CopilotStatus>(`/api/v1/intelligence/copilot/status?tenant_id=${tenantId}`),
    ...SHARED,
  });
}

export function askCopilot(tenantId: string, message: string, conversation: CopilotTurn[], conversationId: string | null): Promise<CopilotResponse> {
  return apiRequest<CopilotResponse>(`/api/v1/intelligence/copilot/query?tenant_id=${tenantId}`, {
    method: "POST",
    body: JSON.stringify({ message, conversation, ...(conversationId ? { conversation_id: conversationId } : {}) }),
  });
}

/**
 * One explanation per exact context (business, kind, record or period, language). It is fetched
 * only when the owner asks (`asked`), never refetched in the background and kept for the session,
 * so reopening the same record shows the same wording until the owner asks again; a different
 * context is a different key and starts unasked, so no summary is ever shown for another context.
 */
export function useExplanation(tenantId: string, context: ExplainContext, language: "en" | "ar", asked: boolean) {
  return useQuery({
    queryKey: [...intelligenceKeys.all(tenantId), "explain", language, context],
    queryFn: () => apiRequest<ExplanationResponse>(`/api/v1/intelligence/explain?tenant_id=${tenantId}`, {
      method: "POST",
      body: JSON.stringify({ ...context, language }),
    }),
    enabled: asked,
    retry: false,
    staleTime: Infinity,
    gcTime: 30 * 60_000,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
  });
}

File: apps/api/tawzeevo_api/schemas/intelligence.py
"""Owner-only intelligence API shapes (D-089). Every money figure travels with its currency, lists
are grouped per currency and nothing is summed across currencies. Scores and bands are workflow
heuristics, never probabilities; cash-flow carries no forecast field."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from tawzeevo_api.schemas.analytics import PeriodInfo

Scalar = Decimal | int | str | bool | None


class Reason(BaseModel):
    code: str
    value: Scalar = None
    context: dict[str, Scalar] = {}


class PriorityComponents(BaseModel):
    """Points earned per component; null = inputs unavailable (excluded from the denominator)."""

    collection_urgency: int | None
    relationship_inactivity: int | None
    activity_decline: int | None
    friction_signals: int | None


class PriorityItem(BaseModel):
    customer_id: UUID
    customer_name: str
    customer_grade: str | None
    currency: str
    score: int
    band: str
    components: PriorityComponents
    reasons: list[Reason]
    suggested_action_code: str
    inactivity_status: str
    outstanding_balance: Decimal
    days_since_last_purchase: int | None


class PriorityGroup(BaseModel):
    currency: str
    items: list[PriorityItem]


class PrioritiesResponse(BaseModel):
    as_of: datetime
    groups: list[PriorityGroup]


class InactivityItem(BaseModel):
    customer_id: UUID
    customer_name: str
    customer_grade: str | None
    currency: str
    status: str
    recency_ratio: Decimal | None
    days_since_last_purchase: int | None
    median_purchase_interval_days: Decimal | None
    purchase_interval_mad_days: Decimal | None
    invoice_count_lifetime: int
    invoice_count_90d: int
    sales_30d: Decimal
    sales_90d: Decimal
    outstanding_balance: Decimal
    last_purchase_at: datetime | None
    reason_codes: list[str]


class InactivityGroup(BaseModel):
    currency: str
    items: list[InactivityItem]


class InactivityResponse(BaseModel):
    as_of: datetime
    groups: list[InactivityGroup]


class AnomalyBaseline(BaseModel):
    method: str
    median: Decimal | None
    mad: Decimal | None
    sample_size: int | None


class AnomalyItem(BaseModel):
    type: str
    severity: str
    detected_at: datetime
    subject_type: str
    subject_id: UUID | None
    metric: str
    observed_value: Decimal
    baseline: AnomalyBaseline
    reason_code: str
    details: dict[str, Scalar]


class AnomalyGroup(BaseModel):
    currency: str
    items: list[AnomalyItem]
    insufficient_history: list[str]


class AnomalyWindow(BaseModel):
    start: datetime
    end: datetime
    timezone: str
    block_days: int
    baseline_blocks: int


class AnomaliesResponse(BaseModel):
    as_of: datetime
    window: AnomalyWindow
    groups: list[AnomalyGroup]


class CashPosition(BaseModel):
    customer_receivables: Decimal
    customer_credit: Decimal
    overdue_receivables: Decimal
    overdue_customer_count: int
    supplier_payables: Decimal
    supplier_credit: Decimal


class AgeingBucket(BaseModel):
    bucket: str
    amount: Decimal
    customer_count: int


class HistoricalFlow(BaseModel):
    """What already happened in the period — history, not a prediction."""

    customer_receipts: Decimal
    customer_refunds: Decimal
    supplier_payments: Decimal
    net_customer_collections: Decimal
    average_weekly_collections: Decimal | None


class PlannedCollections(BaseModel):
    """Open delivery tasks' `amount_to_collect` — a projection that ignores invoice adjustments."""

    amount: Decimal
    task_count: int
    from_date: date
    through_date: date
    source: str
    projection_warning_code: str


class CurrencyCashFlow(BaseModel):
    currency: str
    position: CashPosition
    ageing: list[AgeingBucket]
    historical_flow: HistoricalFlow
    planned_collections: PlannedCollections


class CashFlowResponse(BaseModel):
    as_of: datetime
    period: PeriodInfo
    overdue_threshold_days: int | None
    currencies: list[CurrencyCashFlow]


class CopilotMessage(BaseModel):
    model_config = ConfigDict(extra="forbid")

    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=4000)


class CopilotQueryRequest(BaseModel):
    """One question plus the client-held, bounded history (never stored on the server). The
    history should carry each earlier answer's `conversation_text` (references, not names)."""

    model_config = ConfigDict(extra="forbid")

    message: str = Field(min_length=1, max_length=1000)
    conversation: list[CopilotMessage] = Field(default_factory=list, max_length=12)
    # Echo the previous response's id to continue a conversation; omit it to start a new one
    # (every customer then gets a fresh reference).
    conversation_id: UUID | None = None


class CopilotReference(BaseModel):
    ref: str
    customer_id: UUID
    customer_name: str


class CopilotGrounding(BaseModel):
    tool: str
    period: str | None
    currency: str | None
    ok: bool


class CopilotResponse(BaseModel):
    conversation_id: UUID
    answer: str
    conversation_text: str
    references: list[CopilotReference]
    grounding: list[CopilotGrounding]
    warnings: list[str]
    unverified_numbers: list[str]


class _ExplainRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    language: Literal["en", "ar"] = "en"


class CustomerExplainRequest(_ExplainRequest):
    """A brief about one customer of the owner's business (another tenant's id is not found)."""

    kind: Literal["customer"]
    customer_id: UUID


class AnomalyExplainRequest(_ExplainRequest):
    """One unusual change as the client shows it: its position in the currency's list plus its
    type and subject, re-checked on the server (anomalies are computed, they have no id)."""

    kind: Literal["anomaly"]
    currency: str = Field(pattern=r"^[A-Z]{3}$")
    index: int = Field(ge=0, le=500)
    type: Literal[
        "SALES_PERIOD_HIGH",
        "SALES_PERIOD_LOW",
        "CANCELLATION_SPIKE",
        "REVERSAL_SPIKE",
        "REFUND_SPIKE",
        "CUSTOMER_INVOICE_VALUE_HIGH",
        "BACKDATED_RECEIPT_LARGE",
        "OVERDUE_THRESHOLD_CROSSED",
        "SUPPLIER_PAYABLE_JUMP",
        "LINE_PRICE_BELOW_SNAPSHOT_COST",
    ]
    subject_id: UUID | None = None


class CashExplainRequest(_ExplainRequest):
    """The cash position for the period (and currency) the Analytics screen shows."""

    kind: Literal["cash"]
    period: Literal["30d", "90d", "1y", "all"] = "90d"
    currency: str | None = Field(default=None, pattern=r"^[A-Z]{3}$")


ExplainRequest = Annotated[
    CustomerExplainRequest | AnomalyExplainRequest | CashExplainRequest,
    Field(discriminator="kind"),
]


class ExplanationResponse(BaseModel):
    kind: str
    as_of: datetime
    answer: str
    references: list[CopilotReference]
    grounding: list[CopilotGrounding]
    warnings: list[str]
    unverified_numbers: list[str]


class CopilotStatus(BaseModel):
    configured: bool
    provider: str | None
    model: str | None
